src/backend/distributed/planner/insert_select_planner.c

```diff
@@ -0,0 +1,234 @@
+
+	/* add casts when the SELECT output does not directly match the table */
+	forboth(insertTargetCell, insertSelectQuery->targetList,
+			selectTargetCell, subquery->targetList)
+	{
+		TargetEntry *insertTargetEntry = (TargetEntry *) lfirst(insertTargetCell);
+		TargetEntry *selectTargetEntry = (TargetEntry *) lfirst(selectTargetCell);
+
+		Var *columnVar = NULL;
+		Oid columnType = InvalidOid;
+		int32 columnTypeMod = 0;
+		Oid selectOutputType = InvalidOid;
+
+		/* indirection is not supported, e.g. INSERT INTO table (composite_column.x) */
+		if (!IsA(insertTargetEntry->expr, Var))
+		{
+			ereport(ERROR, (errmsg("can only handle regular columns in the target "
+								   "list")));
+		}
+
+		columnVar = (Var *) insertTargetEntry->expr;
+		columnType = get_atttype(targetRelationId, columnVar->varattno);
+		columnTypeMod = get_atttypmod(targetRelationId, columnVar->varattno);
+		selectOutputType = columnVar->vartype;
+
+		/*
+		 * If the type in the target list does not match the type of the column,
+		 * we need to cast to the column type. PostgreSQL would do this
+		 * automatically during the insert, but we're passing the SELECT
+		 * output directly to COPY.
+		 */
+		if (columnType != selectOutputType)
+		{
+			Expr *selectExpression = selectTargetEntry->expr;
+			Expr *typeCast = (Expr *) coerce_to_target_type(NULL,
+															(Node *) selectExpression,
+															selectOutputType,
+															columnType,
+															columnTypeMod,
+															COERCION_EXPLICIT,
+															COERCE_IMPLICIT_CAST,
+															-1);
+
+			selectTargetEntry->expr = typeCast;
+		}
+	}
+
+	multiPlan->insertSelectQuery = subquery;
+	multiPlan->insertTargetList = insertSelectQuery->targetList;
+	multiPlan->targetRelationId = targetRelationId;
+
+	return multiPlan;
+}
+
+
+/*
+ * ErrorIfCoordinatorInsertSelectUnsupported returns an error if executing an
+ * INSERT ... SELECT command by pulling results of the SELECT to the coordinator
+ * is unsupported because it uses RETURNING, ON CONFLICT, or an append-distributed
+ * table.
+ */
+static DeferredErrorMessage *
+ErrorIfCoordinatorInsertSelectUnsupported(Query *insertSelectQuery)
+{
+	RangeTblEntry *insertRte = NULL;
+
+	if (list_length(insertSelectQuery->returningList) > 0)
+	{
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "RETURNING is not supported in INSERT ... SELECT via "
+							 "coordinator", NULL, NULL);
+	}
+
+	if (insertSelectQuery->onConflict)
+	{
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "ON CONFLICT is not supported in INSERT ... SELECT via "
+							 "coordinator", NULL, NULL);
+	}
+
+	insertRte = ExtractInsertRangeTableEntry(insertSelectQuery);
+	if (PartitionMethod(insertRte->relid) == DISTRIBUTE_BY_APPEND)
+	{
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "INSERT ... SELECT into an append-distributed table is "
+							 "not supported", NULL, NULL);
+	}
+
+	return NULL;
```

Comment:
I think we need to error out if the target table is a reference table and the node is not the coordinator node.